cpinitiative/usaco-guide
# Problems refactor

## Remaining Todos

- [ ] document changes (todo: tomorrow)
- [ ] improve editor experience (ie. don't just crash the moment problem JSON is malformatted) (todo: tomorrow)
- [x] test stuff
- [x] implement redirects
- [x] update liveupdate branding and links to editor
- [x] fix tests ~~lol or just remove them~~
- [x] update problem IDs for user problem solutions

## Changes

(oops these should all be better documented before merging)

- changes problem unique ID from URL to a different string (something like `cf-1137D` for example)
- separates problem definition from the module markdown (so now each module has a JSON file as well as a mdx file)
  - Note that there's a small bug: when you make a change to the JSON file locally, you need to also make a change to the mdx  file (such as adding a random newline somewhere and saving, then deleting the new line) to tell Gatsby to recompile the file
- Rewrites the Problem class to become a plain javascript object; no more magic ID generation stuff
  - Also adds a `ProblemMetadata` class that Gatsby uses to generate `ProblemInfo`
- renames liveupdate to editor (prepping for editor overhaul)
- updates internal solution path to become `/problems/[problem-slug]/solution` (we redirect old links). The slug is generated by slugifying the problem name.
- adds `content/extraProblems.json` -- this file holds problems that don't appear in any module, but appear in search results and the USACO division table.
  - Note that internal solutions MUST be linked to a problem, so if there's an internal solution written for a problem not present in any module, the problem info must be added to the `extraProblems.json` file.
  - Also, if a problem is in this file, it can't be in any other module (if it's in another module and in this file just remove it from this file)
- Forces problems with the same unique ID to have the same source/url/name
- Forces problems to link to internal solutions if those exist
- Internal solution IDs are now changed to be the same as the problem ID that it's a solution for
  - However, you still need to manually go to the problem JSON and specify that an internal solution exists (this will eventually be automated by the new editor)
- Fixes #984 -- search modules now update correctly

## How to Test

Major breaking change, testing help is needed!

1. Back up your data to a JSON file on usaco.guide (very important -- this refactor will delete old data)
2. Go to https://usaco-guide-git-problems-refactor-cpinitiative.vercel.app/ and play around with stuff (especially stuff relating to problems)
3. Go to https://usaco-guide-git-problems-refactor-cpinitiative.vercel.app/ and import your backed-up JSON file and make sure that all the problem completion tracking stuff was migrated correctly
4. Sign in with your main account at https://usaco-guide-git-problems-refactor-cpinitiative.vercel.app/ and make sure stuff was migrated correctly (make sure you've backed up your data before you do this -- this will delete your old data)

## Long-Term Vision

Ideally we hope to significantly improve the functionality of `liveupdate` so that all content-related work can be done 100% online without having to do anything locally. This includes:

- making changes to a module
- adding/removing problems from a module (we'll have a proper UI for this so you shouldn't have to edit the problem JSON file)
- adding internal solutions for a problem (ideally any JSON that refers to this problem will be updated to reflect the new internal solution so you shouldn't have to edit any problem JSON files here either)
- etc etc

---

- Closes #930 
- Closes #984
